File: 210-find-order/Solution.java
import java.util.*;

class Solution210 {
    boolean hasCycle = false;
    Deque<Integer> stack = new ArrayDeque<>();
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] color = new int[numCourses];
        List<List<Integer>> edges = new ArrayList<>();
        for (int i = 0; i < numCourses; i++)
            edges.add(new ArrayList<>());
        for (int[] prereq : prerequisites) //prerequisites 轉鄰接表
        {
            int course = prereq[0]; //想修的課
            int pre = prereq[1]; //先修的課
            edges.get(pre).add(course);
        }   
        for (int i = 0; i < numCourses; i++){
            if (color[i] == 0)
                dfs(edges, color, i);
            if (hasCycle) //如果剛的dfs發現hasCycle, 那後面的也都不用看了, 直接return
                return new int[0];
        }
        int n = stack.size();
        int[] result = new int[n];
        for (int i = 0; i < n; i++) //stack轉成陣列後再return
            result[i] = stack.pollLast();
        return result;
    }
    void dfs(List<List<Integer>> edge, int[] color, int cur){
        if (color[cur] == 1) //走到訪問中的節點, 發現環
        {
            hasCycle = true;
            return;
        }
        if (color[cur] == 2) //走到已訪問的節點, 直接返回
            return;

        color[cur] = 1; //沒被訪問過, 標記為訪問中
        for(int next : edge.get(cur)){
            dfs(edge, color, next);
            if (hasCycle)
                return;
        }
        color[cur] = 2;
        stack.offerLast(cur);
    }
    public static void main(String[] args) {
        Solution210 sol = new Solution210();
//        int[][] prerequisites = {{1,0},{2,0},{3,1},{3,2}};
        int[][] prerequisites = {{1,0},{0,1}};
        System.out.println("Ans = " + Arrays.toString(sol.findOrder(4, prerequisites)));
    }
}

File: 210-find-order/solution.py
from typing import List
from collections import deque
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        def dfs(graph, color, cur):
            if color[cur] != 0:
                return color[cur] == 2
            color[cur] = 1
            for x in graph[cur]:
                if not dfs(graph, color, x):
                    return False
            color[cur] = 2
            temp.appendleft(cur)
            return True

        graph = []
        n = len(prerequisites)
        for _ in range(numCourses):
            graph.append([])
        for i in range(n):
            pre = prerequisites[i][1]
            cur = prerequisites[i][0]
            graph[pre].append(cur)

        color = [0] * numCourses
        temp = deque()
        for i in range (numCourses):
            if color[i] == 0:
                if not dfs(graph, color, i):
                    return []

        result = [0] * numCourses
        for i in range(len(temp)):
            result[i] = temp[i]
        return result

if __name__ == "__main__":
    # prerequisites = [[1, 0], [2, 0], [3, 1], [3, 2]]
    prerequisites = []
    print("Ans = ", Solution().findOrder(2, prerequisites))

"""
寫第二遍, 自己寫出來, 做了代碼優化

相似題:
207, 210, 785, 802
"""